File: Sección02_Regresión/Regresion Lineal Multiple/multiple_linear_regression.R
# Regresión Lineal Múltiple
# Importar el dataset
dataset = read.csv('50_Startups.csv')
# dataset = dataset[,2:3]

# Codificar las variables categóricas
dataset$State = factor(dataset$State,
                         levels = c("New York", "California", "Florida"),
                         labels = c(1,2,3))

# Dividir data set en conjunto de entrenamiento y testing
# install.packages("caTools")
library(caTools)
set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
testing_set = subset(dataset, split == FALSE)

# Escalado de valores
# training_set[,2:3] = scale(training_set[,2:3])
# testing_set[,2:3] = scale(testing_set[,2:3])

# Ajustar el modelo de Regresión Lineal Múltiple al conjunto de Entrenamiento
regression = lm(formula = Profit ~ .,
                data = training_set)

# Predecir los resultados con el conjunto de Testing
y_pred = predict(regression, newdata = testing_set)

# Construir un modelo óptimo con la eliminación hacia atrás
SL = 0.05
regression = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State ,
                data = dataset)

regression = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend ,
                data = dataset)

regression = lm(formula = Profit ~ R.D.Spend + Marketing.Spend ,
                data = dataset)

regression = lm(formula = Profit ~ R.D.Spend ,
                data = dataset)

summary(regression)

# install.packages("https://cran.r-project.org/src/contrib/Archive/ElemStatLearn/ElemStatLearn_2015.6.26.2.tar.gz",repos=NULL, type="source")

File: Sección02_Regresión/Regresion Lineal Multiple/multiple_linear_regression.py
# -*- coding: utf-8 -*-
"""
Created on Mon Mar 27 15:19:13 2023

@author: Ckonny
"""

# Regresión Lineal Múltiple

# Importar las librerías
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importar el dataset
dataset = pd.read_csv('50_Startups.csv')
X =     dataset.iloc[:,:-1].values
y =     dataset.iloc[:,4].values

# Codificar datos Categóricos
from sklearn import preprocessing
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

le_X = preprocessing.LabelEncoder()
X[:,3] = le_X.fit_transform(X[:,3])
ct = ColumnTransformer(
    [('one_hot_encoder', OneHotEncoder(categories='auto'), [3])],   
    remainder='passthrough')
X = np.array(ct.fit_transform(X), dtype=float)
# En consola para la visualización de decimales
# np.set_printoptions(suppress=True)

# Evitar la trampa de las variables ficticias (eliminamos 1)
X = X[:, 1:]

# Dividir el dataset en conjunto de entrenamiento y testing
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.2, random_state = 0)

# Escalado de Variable
""" from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test) """

# Ajustar el modelo de Regresión Lineal Múltiple  con el conjunto de entrenamiento
from sklearn.linear_model import LinearRegression
regression = LinearRegression()
regression.fit(X_train, y_train)

# Predicción de los resultados en el conjunto de Testing
y_pred = regression.predict(X_test)

# Construir el modelo óptimo de RLM utilizando la eliminación hacia atrás
import statsmodels.formula.api as sm
import statsmodels.api as SM
X = np.append(arr = np.ones((50,1)).astype(int), values = X, axis = 1)
SL = 0.05

X_opt = X[:, [0, 1, 2, 3, 4, 5]]
regressor_OLS = SM.OLS(y, X_opt).fit()
regressor_OLS.summary()

X_opt = X[:, [0, 1, 3, 4, 5]]
regressor_OLS = SM.OLS(y, X_opt).fit()
regressor_OLS.summary()

X_opt = X[:, [0, 3, 4, 5]]
regressor_OLS = SM.OLS(y, X_opt).fit()
regressor_OLS.summary()

X_opt = X[:, [0, 3, 5]]
regressor_OLS = SM.OLS(y, X_opt).fit()
regressor_OLS.summary()

X_opt = X[:, [0, 3]]
regressor_OLS = SM.OLS(y, X_opt).fit()
regressor_OLS.summary()